### ISSUE TITLE
A complete PyTorch reproduction of CliffordNet

### ISSUE THREAD

**Issue by chenziwenhaoshuai** [ParaMind2025/CAN#4]
Hi, thanks for your excellent work on *Clifford Neural Networks for Vision*.  
Since the official code has not yet been released, I spent some time reproducing the model and training pipeline in PyTorch.

I would like to share my implementation here in case it is useful for other researchers who want to run experiments before the official release.

🔗 **Repository:**  
https://github.com/chenziwenhaoshuai/CliffordNet-pytorch

### Key Performance Metrics
| Metric | Nano               | Fast  | Base  |
|--------|--------------------|-------|-------|
| Best Test Accuracy | **77.94%** (Epoch 192) | **81.18%**  (Epoch 199)   | **81.95%**  (Epoch 200)     |
| Results in the paper | 76.41%             | 77.63% | 78.05% |

### ✔ Reproducibility
The implementation adheres closely to the architecture and algorithmic details described in the paper.  
With the full training setup in place, we were able to obtain results that are fully reproducible — and in our tests, the models even performed slightly above what was originally reported.

If any architectural detail requires adjustment or if an official implementation is planned, I’m happy to update or align the code accordingly.

Thanks again for the great work!

**Comment by ychuest** [comment]
> Hi, thanks for your excellent work on _Clifford Neural Networks for Vision_. Since the official code has not yet been released, I spent some time reproducing the model and training pipeline in PyTorch.
> 
> I would like to share my implementation here in case it is useful for other researchers who want to run experiments before the official release.
> 
> 🔗 **Repository:** https://github.com/chenziwenhaoshuai/CliffordNet-pytorch
> 
> ### ✔ Reproducibility
> The implementation adheres closely to the architecture and algorithmic details described in the paper. With the full training setup in place, we were able to obtain results that are fully reproducible — and in our tests, the models even performed slightly above what was originally reported.
> 
> If any architectural detail requires adjustment or if an official implementation is planned, I’m happy to update or align the code accordingly.
> 
> Thanks again for the great work!

wonderful！！！！

**Comment by zikai1** [comment]
Great!

**Comment by ZhangQuan-hub** [comment]
excellent job!

**Comment by jackalcooper** [comment]
People soon would find this to be the true way towards AGI.

**Comment by johndpope** [comment]
Hey @ParaMind2025  - i took the clifford algebra - inspired from your paper - and upgraded TurboQuant to use them 

https://www.reddit.com/r/LocalLLaMA/comments/1s44p77/rotorquant_1019x_faster_alternative_to_turboquant/

**Comment by ParaMind2025** [comment]
> Hey [@ParaMind2025](https://github.com/ParaMind2025) - i took the clifford algebra - inspired from your paper - and upgraded TurboQuant to use them
> 
> https://www.reddit.com/r/LocalLLaMA/comments/1s44p77/rotorquant_1019x_faster_alternative_to_turboquant/

@johndpope Hi, thank you for your message and your excellent work.
I made a simple extension based on your work, incorporating some of the techniques from our this paper https://arxiv.org/abs/2506.07405.
However, it is still just a preliminary experiment at this stage.
https://github.com/ParaMind2025/isoquant

**Comment by johndpope** [comment]
its working - I upstreamed to repo - it's 5x - well done. 
https://github.com/scrya-com/rotorquant/commit/f246855064798d07539ee6d29d0d8aa03ae25bf3

UPDATE

when i run this with the triton kernel - it's only 4 nanoseconds faster..
but i guess the memory is happier.
<img width="951" height="568" alt="Image" src="https://github.com/user-attachments/assets/0d6e0e5f-9374-4e8c-98f0-aebc131a84ff" />


I implement isoquant with triton kernel - so it works with amd too - 
https://github.com/scrya-com/rotorquant
i also add some benchmarks -

**Comment by ParaMind2025** [comment]
> its working - I upstreamed to repo - it's 5x - well done. [scrya-com/rotorquant@f246855](https://github.com/scrya-com/rotorquant/commit/f246855064798d07539ee6d29d0d8aa03ae25bf3)
> 
> UPDATE
> 
> when i run this with the triton kernel - it's only 4 nanoseconds faster.. but i guess the memory is happier. <img alt="Image" width="951" height="568" src="https://private-user-images.githubusercontent.com/289994/571051976-0d6e0e5f-9374-4e8c-98f0-aebc131a84ff.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQ4NTU5NTksIm5iZiI6MTc3NDg1NTY1OSwicGF0aCI6Ii8yODk5OTQvNTcxMDUxOTc2LTBkNmUwZTVmLTkzNzQtNGU4Yy05OGYwLWFlYmMxMzFhODRmZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMzMFQwNzI3MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNzRkZjIyMjMzNjQ3MDIxM2IxN2Y2NzEyZmNlMDU4YTA5Y2U2MWY3ZjJkMTk3OGExZjY3OGFjOWU3ZTA4Mzc3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.c9rtMuI8TLmeNEHo8lqckESHi38DUj5It8o1dGa7nno">
> 
> I implement isoquant with triton kernel - so it works with amd too - https://github.com/scrya-com/rotorquant i also add some benchmarks - 

@johndpope  Thank you for your work. I also added a 2D version as a special case of SO(4), which delivers performance roughly on par with the IsoQuant-Fast variant. https://github.com/ParaMind2025/isoquant/blob/main/isoquant/csrc/planar2_fused_kernel.cu

**Comment by johndpope** [comment]
Hey Zhongping,  i upstream that -   PlanarQuant -
 at the triton level the speeds differences are invisible.
